File: src/underwright/application/services/customer_profile_service.py
from __future__ import annotations

from underwright.domain.auth_user import AuthUser
from underwright.domain.customer_profile import (
    CustomerAddressProfile,
    CustomerProfileCompletionSource,
    CustomerProfileReadModel,
    CustomerProfileStatus,
    CustomerProfileUpdate,
    StoredCustomerProfile,
)


class CustomerProfileError(ValueError):
    pass


class CustomerProfileIncompleteError(CustomerProfileError):
    def __init__(
        self,
        message: str = "Customer profile is incomplete.",
        *,
        missing_fields: list[str] | None = None,
        status: CustomerProfileStatus = "incomplete",
    ) -> None:
        super().__init__(message)
        self.missing_fields = missing_fields or []
        self.status = status


class CustomerProfileNotFoundError(CustomerProfileError):
    pass


class CustomerProfileService:
    """Owns profile completeness for the auth_user.client_id customer link."""

    def __init__(self, customer_profile_repository, auth_user_repository) -> None:
        self.customer_profile_repository = customer_profile_repository
        self.auth_user_repository = auth_user_repository

    def get_profile(self, user: AuthUser) -> CustomerProfileReadModel:
        if user.role != "client":
            raise CustomerProfileNotFoundError("Customer profile is client-only.")
        if user.client_id is None:
            return self._pending_profile(user)
        return self.get_customer_profile(user.client_id)

    def get_customer_profile(self, customer_id: int) -> CustomerProfileReadModel:
        try:
            stored = self.customer_profile_repository.get_customer_profile(customer_id)
        except ValueError as exc:
            raise CustomerProfileNotFoundError("Customer profile not found.") from exc
        linked_count = self._linked_auth_user_count(customer_id)
        return self._read_model(stored, linked_auth_user_count=linked_count)

    def list_customer_profiles(self) -> list[CustomerProfileReadModel]:
        profiles = self.customer_profile_repository.list_customer_profiles()
        return [
            self._read_model(
                profile,
                linked_auth_user_count=self._linked_auth_user_count(
                    profile.customer_id
                ),
            )
            for profile in profiles
        ]

    def update_profile(
        self,
        *,
        user: AuthUser,
        update: CustomerProfileUpdate,
        source: CustomerProfileCompletionSource = "client_self_service",
    ) -> CustomerProfileReadModel:
        existing = self._stored_profile_or_none(user)
        draft = self._merge_profile(user=user, existing=existing, update=update)
        missing_fields = self._missing_fields(draft)
        if missing_fields:
            raise CustomerProfileIncompleteError(
                missing_fields=missing_fields,
                status="pending_customer_link" if user.client_id is None else "incomplete",
            )

        complete_profile = StoredCustomerProfile.model_validate(draft)
        if user.client_id is None:
            stored = self.customer_profile_repository.create_customer_profile(
                complete_profile,
                updated_by_auth_user_id=user.id,
                source=source,
            )
            self.auth_user_repository.update_user_client_id(
                user_id=user.id,
                client_id=stored.customer_id,
            )
        else:
            stored = self.customer_profile_repository.update_customer_profile(
                customer_id=user.client_id,
                profile=complete_profile,
                status="complete",
                updated_by_auth_user_id=user.id,
                source=source,
            )
        return self._read_model(stored, linked_auth_user_count=1)

    def ensure_complete_profile(self, user: AuthUser) -> CustomerProfileReadModel:
        profile = self.get_profile(user)
        if profile.status != "complete":
            raise CustomerProfileIncompleteError(
                missing_fields=profile.missing_fields,
                status=profile.status,
            )
        return profile

    def profile_status_for_user(self, user: AuthUser) -> CustomerProfileStatus | None:
        if user.role != "client":
            return None
        return self.get_profile(user).status

    def mark_employee_link(
        self,
        *,
        customer_id: int,
        updated_by_auth_user_id: int | None,
    ) -> CustomerProfileReadModel:
        stored = self.customer_profile_repository.get_customer_profile(customer_id)
        status = "complete" if not self._missing_fields(stored.model_dump()) else "incomplete"
        updated = self.customer_profile_repository.touch_customer_profile(
            customer_id=customer_id,
            status=status,
            updated_by_auth_user_id=updated_by_auth_user_id,
            source="employee_link",
        )
        return self._read_model(updated)

    def _stored_profile_or_none(
        self,
        user: AuthUser,
    ) -> StoredCustomerProfile | None:
        if user.client_id is None:
            return None
        try:
            return self.customer_profile_repository.get_customer_profile(user.client_id)
        except ValueError as exc:
            raise CustomerProfileNotFoundError("Customer profile not found.") from exc

    def _pending_profile(self, user: AuthUser) -> CustomerProfileReadModel:
        draft = {
            "customer_id": None,
            "type": None,
            "full_name": user.full_name,
            "national_id": None,
            "company_id": None,
            "email": user.email,
            "phone": user.phone,
            "address": None,
        }
        return CustomerProfileReadModel(
            **draft,
            status="pending_customer_link",
            requires_customer_profile_completion=True,
            missing_fields=self._missing_fields(draft),
        )

    def _read_model(
        self,
        stored: StoredCustomerProfile,
        *,
        linked_auth_user_count: int | None = None,
    ) -> CustomerProfileReadModel:
        missing_fields = self._missing_fields(stored.model_dump())
        status: CustomerProfileStatus = "complete" if not missing_fields else "incomplete"
        return CustomerProfileReadModel(
            customer_id=stored.customer_id,
            status=status,
            requires_customer_profile_completion=status != "complete",
            type=stored.type,
            full_name=stored.full_name,
            national_id=stored.national_id,
            company_id=stored.company_id,
            email=stored.email,
            phone=stored.phone,
            address=stored.address,
            missing_fields=missing_fields,
            customer_profile_completed_at=stored.customer_profile_completed_at,
            customer_profile_updated_at=stored.customer_profile_updated_at,
            customer_profile_updated_by_auth_user_id=(
                stored.customer_profile_updated_by_auth_user_id
            ),
            customer_profile_completion_source=(
                stored.customer_profile_completion_source
            ),
            profile_update_count=stored.profile_update_count,
            linked_auth_user_count=linked_auth_user_count,
        )

    def _merge_profile(
        self,
        *,
        user: AuthUser,
        existing: StoredCustomerProfile | None,
        update: CustomerProfileUpdate,
    ) -> dict:
        base = (
            existing.model_dump()
            if existing is not None
            else {
                "customer_id": 0,
                "type": None,
                "full_name": user.full_name,
                "national_id": None,
                "company_id": None,
                "email": user.email,
                "phone": user.phone,
                "address": {},
            }
        )
        update_values = update.model_dump(exclude_unset=True)
        address_update = update_values.pop("address", None)
        for field_name, value in update_values.items():
            base[field_name] = self._clean(value)

        address = dict(base.get("address") or {})
        if isinstance(base.get("address"), CustomerAddressProfile):
            address = base["address"].model_dump()
        if address_update is not None:
            for field_name, value in address_update.items():
                address[field_name] = self._clean(value)
        base["address"] = self._address_with_full_text(address)
        return base

    def _address_with_full_text(self, address: dict) -> dict:
        full_text = self._clean(address.get("full_text"))
        if not full_text:
            parts = [
                address.get("street"),
                address.get("number"),
                address.get("city"),
                address.get("county"),
                address.get("country"),
                address.get("postal_code"),
            ]
            full_text = ", ".join(
                self._clean(part) for part in parts if self._clean(part)
            )
        address["full_text"] = full_text
        return address

    def _missing_fields(self, profile: dict) -> list[str]:
        missing: list[str] = []
        customer_type = profile.get("type")
        for field_name in ("type", "full_name", "email", "phone"):
            if not self._present(profile.get(field_name)):
                missing.append(field_name)

        if customer_type == "individual" and not self._present(
            profile.get("national_id")
        ):
            missing.append("national_id")
        if customer_type == "company" and not self._present(profile.get("company_id")):
            missing.append("company_id")

        address = profile.get("address")
        if isinstance(address, CustomerAddressProfile):
            address = address.model_dump()
        if not isinstance(address, dict):
            address = {}
        for field_name in (
            "country",
            "county",
            "city",
            "street",
            "number",
            "postal_code",
        ):
            if not self._present(address.get(field_name)):
                missing.append(f"address.{field_name}")
        return missing

    def _linked_auth_user_count(self, customer_id: int) -> int:
        if not hasattr(self.auth_user_repository, "list_users_by_client_id"):
            return 0
        return len(self.auth_user_repository.list_users_by_client_id(customer_id))

    def _present(self, value) -> bool:
        return bool(self._clean(value))

    def _clean(self, value) -> str | None:
        if value is None:
            return None
        if not isinstance(value, str):
            return str(value)
        stripped = value.strip()
        return stripped or None


__all__ = [
    "CustomerProfileError",
    "CustomerProfileIncompleteError",
    "CustomerProfileNotFoundError",
    "CustomerProfileService",
]

File: frontend/src/services/mock/customerProfileService.ts
import { mockUsers } from "../../data/mockUsers";
import type { AppUser, CustomerLegalDocument, CustomerProfile, CustomerProfileUpdate } from "../../types";
import { readStoredAuthUser, setAuthSession } from "../authSession";
import { delay, readStored, writeStored } from "../storage";

const usersKey = "ultrasafe_users_v4";

export async function getMyCustomerProfile(): Promise<CustomerProfile> {
  return delay(toCustomerProfile(readStoredAuthUser()));
}

export async function updateMyCustomerProfile(
  profile: CustomerProfileUpdate
): Promise<CustomerProfile> {
  const user = readStoredAuthUser();
  if (!user) throw new Error("Sign in is required.");
  const missing = missingFields(profile, user);
  if (missing.length) {
    throw new Error(`CUSTOMER_PROFILE_INCOMPLETE: ${missing.join(", ")}`);
  }
  const customerId = user.customerId ?? String(Date.now());
  const updated: AppUser = {
    ...user,
    id: customerId,
    customerId,
    customerProfileStatus: "complete",
    requiresCustomerProfileCompletion: false,
    fullName: profile.full_name ?? user.fullName,
    email: profile.email ?? user.email,
    phone: profile.phone ?? user.phone,
    nationalId: profile.national_id ?? user.nationalId,
    address: formatAddress(profile.address)
  };
  const users = readStored(usersKey, mockUsers);
  writeStored(
    usersKey,
    users.map((item) => (item.email === user.email ? updated : item))
  );
  if (typeof window !== "undefined") {
    setAuthSession(updated);
  }
  return delay(toCustomerProfile(updated, profile));
}

export async function getCustomerProfiles(): Promise<CustomerProfile[]> {
  const users = readStored(usersKey, mockUsers);
  return delay(
    users.filter((user) => user.role === "client").map((user) => toCustomerProfile(user))
  );
}

function toCustomerProfile(
  user: AppUser | null,
  override?: CustomerProfileUpdate
): CustomerProfile {
  if (!user) {
    return {
      status: "pending_customer_link",
      requires_customer_profile_completion: true,
      missing_fields: ["type", "full_name", "email", "phone", "address.country"]
    };
  }
  const profile: CustomerProfileUpdate = {
    type: override?.type ?? "individual",
    full_name: override?.full_name ?? user.fullName,
    national_id: override?.national_id ?? user.nationalId,
    company_id: override?.company_id,
    email: override?.email ?? user.email,
    phone: override?.phone ?? user.phone,
    address: override?.address ?? parseAddress(user.address)
  };
  const missing = user.requiresCustomerProfileCompletion
    ? missingFields(profile, user)
    : [];
  const complete = Boolean(user.customerId) && !missing.length;
  return {
    customer_id: user.customerId ? Number.parseInt(user.customerId, 10) || null : null,
    status: complete ? "complete" : "pending_customer_link",
    requires_customer_profile_completion: !complete,
    type: profile.type,
    full_name: profile.full_name,
    national_id: profile.national_id,
    company_id: profile.company_id,
    email: profile.email,
    phone: profile.phone,
    address: profile.address,
    missing_fields: complete ? [] : missing,
    customer_profile_completed_at: complete ? new Date().toISOString() : null,
    customer_profile_updated_at: new Date().toISOString(),
    customer_profile_completion_source: complete ? "client_self_service" : null,
    profile_update_count: complete ? 1 : 0,
    linked_auth_user_count: user.customerId ? 1 : 0,
    legal_documents: complete ? mockCustomerLegalDocuments(user, profile) : []
  };
}

function mockCustomerLegalDocuments(
  user: AppUser,
  profile: CustomerProfileUpdate
): CustomerLegalDocument[] {
  const baseId = user.customerId ?? user.id;
  const filePrefix = (profile.full_name ?? user.fullName)
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "") || "customer";
  const documents: CustomerLegalDocument[] = [];

  if (profile.type === "individual" && profile.national_id) {
    documents.push({
      id: `${baseId}-id-document`,
      label: "ID document",
      document_type: "ID document",
      file_name: `${filePrefix}-id-document.pdf`,
      content_type: "application/pdf",
      source: "client_profile",
      extracted_fields: {
        full_name: profile.full_name,
        national_id: profile.national_id
      }
    });
  }

  if (profile.type === "company" && profile.company_id) {
    documents.push({
      id: `${baseId}-company-registration`,
      label: "Company registration document",
      document_type: "Company registration document",
      file_name: `${filePrefix}-company-registration.pdf`,
      content_type: "application/pdf",
      source: "client_profile",
      extracted_fields: {
        company_name: profile.full_name,
        company_id: profile.company_id
      }
    });
  }

  documents.push({
    id: `${baseId}-terms-consent`,
    label: "Terms / consent document",
    document_type: "Terms / consent document",
    file_name: `${filePrefix}-terms-consent.pdf`,
    content_type: "application/pdf",
    source: "client_profile",
    extracted_fields: {
      full_name: profile.full_name,
      email: profile.email
    }
  });

  return documents;
}

function missingFields(profile: CustomerProfileUpdate, user: AppUser) {
  const missing: string[] = [];
  const address = profile.address ?? {};
  if (!profile.type) missing.push("type");
  if (!present(profile.full_name ?? user.fullName)) missing.push("full_name");
  if (!present(profile.email ?? user.email)) missing.push("email");
  if (!present(profile.phone ?? user.phone)) missing.push("phone");
  if (profile.type === "individual" && !present(profile.national_id ?? user.nationalId)) {
    missing.push("national_id");
  }
  if (profile.type === "company" && !present(profile.company_id)) {
    missing.push("company_id");
  }
  for (const field of ["country", "county", "city", "street", "number", "postal_code"] as const) {
    if (!present(address[field])) missing.push(`address.${field}`);
  }
  return missing;
}

function parseAddress(address?: string) {
  return address ? { full_text: address } : undefined;
}

function formatAddress(address?: CustomerProfileUpdate["address"]) {
  if (!address) return "";
  return (
    address.full_text ||
    [
      `${address.street ?? ""} ${address.number ?? ""}`.trim(),
      address.city,
      address.county,
      address.country,
      address.postal_code
    ].filter(Boolean).join(", ")
  );
}

function present(value?: string | null) {
  return Boolean(value?.trim());
}


